src/BuiltInTools/dotnet-watch/Internal/ProcessRunner.cs

```diff
@@ -27,123 +38,135 @@ public static async Task<int> RunAsync(ProcessSpec processSpec, IReporter report
+            processTerminationToken.Register(() => TerminateProcess(process, state, reporter));
 
             stopwatch.Start();
 
-            int? processId = null;
+            Exception? launchException = null;
             try
             {
-                if (process.Start())
+                if (!process.Start())
                 {
-                    processId = process.Id;
+                    throw new InvalidOperationException("Process can't be started.");
                 }
-            }
-            finally
-            {
-                var argsDisplay = processSpec.GetArgumentsDisplay();
 
-                if (processId.HasValue)
-                {
-                    reporter.Report(MessageDescriptor.LaunchedProcess, processSpec.Executable, argsDisplay, processId.Value);
-                }
-                else
+                state.ProcessId = process.Id;
+
+                if (onOutput != null)
                 {
-                    reporter.Error($"Failed to launch '{processSpec.Executable}' with arguments '{argsDisplay}'");
+                    process.BeginOutputReadLine();
+                    process.BeginErrorReadLine();
                 }
             }
+            catch (Exception e)
+            {
+                launchException = e;
+            }
 
-            if (processId == null)
+            var argsDisplay = processSpec.GetArgumentsDisplay();
+            if (launchException == null)
+            {
+                reporter.Report(MessageDescriptor.LaunchedProcess, processSpec.Executable, argsDisplay, state.ProcessId);
+            }
+            else
             {
-                // failed to launch
+                reporter.Error($"Failed to launch '{processSpec.Executable}' with arguments '{argsDisplay}': {launchException.Message}");
                 return int.MinValue;
             }
 
-            if (onOutput != null)
+            if (launchResult != null)
             {
-                process.BeginOutputReadLine();
-                process.BeginErrorReadLine();
+                launchResult.ProcessId = process.Id;
             }
 
             int? exitCode = null;
-            var failed = false;
 
             try
             {
-                await processState.Task;
+                try
+                {
+                    await process.WaitForExitAsync(processTerminationToken);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Process termination requested via cancellation token.
+                    // Wait for the actual process exit.
+                    while (true)
+                    {
+                        // non-cancellable to not leave orphaned processes around blocking resources:
+                        var waitForExit = process.WaitForExitAsync(CancellationToken.None);
+
+                        await Task.WhenAny(waitForExit, Task.Delay(TimeSpan.FromSeconds(5), CancellationToken.None));
```

Comment:
This pattern creates a dangling Timer https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#cancelling-uncancellable-operations.

```suggestion
                        await waitForExit.WaitAsync(TimeSpan.FromSeconds(5));
```

Reply:
Good catch. Thanks!

Reply:
Are you changing it?